x-pack/test/plugin_api_integration/test_suites/task_manager/task_priority.ts

```diff
@@ -0,0 +1,216 @@
+    if (taskId) {
+      filter.push({ term: { taskId } });
+    }
+    if (taskType) {
+      filter.push({ term: { taskType } });
+    }
+    return es
+      .search({
+        index: testHistoryIndex,
+        body: {
+          query: {
+            bool: {
+              filter,
+            },
+          },
+        },
+      })
+      .then((result) => (result as unknown as SearchResults).hits.hits);
+  }
+
+  describe('task priority', () => {
+    beforeEach(async () => {
+      const exists = await es.indices.exists({ index: testHistoryIndex });
+      if (exists) {
+        await es.deleteByQuery({
+          index: testHistoryIndex,
+          refresh: true,
+          body: { query: { term: { type: 'task' } } },
+        });
+      } else {
+        await es.indices.create({
+          index: testHistoryIndex,
+          body: {
+            mappings: {
+              properties: {
+                type: {
+                  type: 'keyword',
+                },
+                taskId: {
+                  type: 'keyword',
+                },
+                params: taskManagerIndexMapping.params,
+                state: taskManagerIndexMapping.state,
+                runAt: taskManagerIndexMapping.runAt,
+              } as Record<string, estypes.MappingProperty>,
+            },
+          },
+        });
+      }
+    });
+
+    afterEach(async () => {
+      await supertest.delete('/api/sample_tasks').set('kbn-xsrf', 'xxx').expect(200);
+    });
+
+    it('should claim low priority tasks if there is capacity', async () => {
+      // schedule 5 normal tasks and 1 low priority task
+      // setting the schedule long so they should only run once
+      const tasksToSchedule = [];
+      for (let i = 0; i < 5; i++) {
+        tasksToSchedule.push(
+          scheduleTask({
+            taskType: 'sampleTask',
+            schedule: { interval: `1d` },
+            params: {},
+          })
+        );
+      }
+      tasksToSchedule.push(
+        scheduleTask({
+          taskType: 'lowPriorityTask',
+          schedule: { interval: `1d` },
+          params: {},
+        })
+      );
+      const scheduledTasks = await Promise.all(tasksToSchedule);
+
+      await retry.try(async () => {
+        const tasks = (await currentTasks()).docs;
+        expect(tasks.length).to.eql(6);
+
+        const taskIds = tasks.map((task) => task.id);
+        const taskDocs: RawDoc[] = [];
+        await asyncForEach(scheduledTasks, async (scheduledTask) => {
+          expect(taskIds).to.contain(scheduledTask.id);
+          const doc: RawDoc[] = await historyDocs({ taskId: scheduledTask.id });
+          expect(doc.length).to.eql(1);
+          taskDocs.push(...doc);
+        });
+
+        expect(
+          taskDocs.findIndex((taskDoc) => taskDoc._source.taskType === 'lowPriorityTask')
+        ).to.be.greaterThan(-1);
+      });
+    });
+
+    it('should not claim low priority tasks when there is no capacity', async () => {
```

Comment:
heh, I was going to ask about this.

Makes sense, except what happens if there's **always** >= WORKER tasks to run of non-low priority, and there's a low priority task queued as well.  Seems like the low priority task will never run.

Likewise, if for some reason there are a number of high priority tasks in the queue, we'd never see mid- or low-priority tasks run.

The main thing I'm thinking about is when we had the issue that tasks that wouldn't run were getting claimed, and so we'd end up getting to the point where that's all that was claimed, and would never run another task.  I think that issue has been resolved, but I'm going to try a few scenarios locally to see if I can break something ...

Reply:
I think having the integration test to let us know if a task is trying to set itself to high priority will help prevent a lot of high priority tasks. If we want to hold off on high priority at all now until we see a need for it, I can remove it from the enum, WDYT?

I can update the README to include a section about priority so that people who use it will be aware that this is a situation that can happen (low priority tasks won't be claimed if there's never any capacity). For the backfill rule runs, we had discussed with Defazio how we might show indicators that backfill tasks aren't getting run due to capacity so we can surface that to the user.

Reply:
Ya, let's remove the high priority constant.

I was able to starve lower-priority tasks by having 100 index threshold rules (changed their priority to 100) running at 1s, and then creating a metric threshold at priority 50.  It never ran :-)

So I think if we keep it to only setting the priority lower (and not higher), we'll only only potentially starve those low priority processes.  Currently it's just the "backfill alerts" tasks we're planning for this, right?  Is there one task to deal with all of them, or one task per rule being back-filled?

I have a feeling we may want to know when this is happening, if we think it might be.  Presumably a query on the TM index could tell us - which is probably good enough for SDH usage.  I wonder if including some "progress" in the backfill status would also (essentially) give us this (as well as potentially to the customers) - perhaps the last date processed for this run, or such.   Or maybe a task manager health status?  I guess we can open an issue if we find this situation, but it ends up being difficult to diagnose.